README.md
<h1>git-server</h1>

A GitHub Protocol & API emulation.

<!-- TOC -->
- [Status](#status)
- [Description](#description)
- [Installation](#installation)
- [Getting started](#getting-started)
  - [1. Create a local Git repository](#1-create-a-local-git-repository)
  - [2. Start server](#2-start-server)
  - [3. Fetch raw content of file in Git repo over http](#3-fetch-raw-content-of-file-in-git-repo-over-http)
- [Git Protocols and APIs](#git-protocols-and-apis)
  - [1. Git Raw Protocol](#1-git-raw-protocol)
  - [2. Git HTTP Transfer Protocols](#2-git-http-transfer-protocols)
    - [`git push` support](#git-push-support)
  - [3. GitHub API v3](#3-github-api-v3)
  - [4. GitHub-like Web Server](#4-github-like-web-server)
<!-- /TOC -->

---
## Status

[![codecov](https://img.shields.io/codecov/c/github/adobe/git-server.svg)](https://codecov.io/gh/adobe/git-server)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/git-server.svg)](https://circleci.com/gh/adobe/git-server)
[![GitHub license](https://img.shields.io/github/license/adobe/git-server.svg)](https://github.com/adobe/git-server/blob/main/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/git-server.svg)](https://github.com/adobe/git-server/issues)
[![Renovate enabled](https://img.shields.io/badge/renovate-enabled-brightgreen.svg)](https://renovatebot.com/)
[![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/git-server.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/git-server/context:javascript)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## Description

`git-server` serves a hierarchy of local Git repositories to Git clients accessing the repository over `http://` and `https://` protocol.

`git-server` expects the local Git repositories (either cloned or created with `git init`) to be organized as follows in the local file system:

```ascii
<repos_root_dir>/
├── <owner_1>/
│   └── <repo_1>
│   └── <repo_2>
├── <owner_2>/
│   └── <repo_1>
│   └── <repo_2>
```

Alternatively, a virtual repository mapping allows to 'mount' repositories independent of their location in the file system into a `<owner>/<repo>` hierarchy. A configuration example:

```js
  virtualRepos: {
    owner1: {
      repo1: {
        path: './repo1',
      },
    },
    owner2: {
      repo2: {
        path: '/repos/repo2',
      },
    },
  },
```

Repositories exposed via `git-server` can be used just like any repository hosted on GitHub, i.e. you can clone them and push changes to.
The repository contents can be accessed with the same url patterns you would use to request files managed on GitHub.

The following protocols and APIs are currently supported:

* [Git Raw protocol](#1-git-raw-protocol)
* [Git HTTP Transfer Protocols](#2-git-http-transfer-protocols)
* [GitHub API v3](#3-github-api-v3)

## Installation

```bash
cd <checkout dir>
npm install
```

## Getting started

`git-server` is configured via the `config.js` file which is expected to exist in the current working directory. Here's the default configuration:

```javascript
{
  appTitle: 'Helix Git Server',
  repoRoot: './repos',
  // repository mapping. allows to 'mount' repositories outside the 'repoRoot' structure.
  virtualRepos: {
    demoOwner: {
      demoRepo: {
        path: './virtual/example',
      },
    },
  },
  listen: {
    http: {
      port: 5000,
      host: '0.0.0.0',
    },
    /*
    // https is optional
    https: {
      // cert: if no file is specfied a selfsigned certificate will be generated on-the-fly
      // cert: './localhost.crt',
      // key: if no file is specfied a key will be generated on-the-fly
      // key: './localhost.key',
      port: 5443,
      host: '0.0.0.0',
    },
    */
  },
  subdomainMapping: {
    // if enabled, <subdomain>.<baseDomain>/foo/bar/baz will be
    // resolved/mapped to 127.0.0.1/<subdomain>/foo/bar/baz
    enable: true,
    baseDomains: [
      // some wildcarded DNS domains resolving to 127.0.0.1
      'localtest.me',
      'lvh.me',
      'vcap.me',
      'lacolhost.com',
    ],
  },
  logs: {
    level: 'info', // fatal, error, warn, info, verbose, debug, trace
    logsDir: './logs',
    reqLogFormat: 'short', // used for morgan (request logging)
  },
}
```

`git-server` uses [helix-log](https://github.com/adobe/helix-log) for logging. The log level (`fatal`, `error`, `warn`, `info`, `verbose`, `debug`, `trace`) can be set via the `logs.level` property in the config (see above). For more information please refer to the [helix-log](https://github.com/adobe/helix-log) project.

### 1. Create a local Git repository

```bash
cd ./repos
# create org
mkdir helix && cd helix
# initialize new git repo
mkdir test && cd test && git init
# allow to remotely push to this repo
git config receive.denyCurrentBranch updateInstead
# add a README.md
echo '# test repo'>README.md
git add . && git commit -m '1st commit'
cd ../../..
```

### 2. Start server

```bash
npm start
```

### 3. Fetch raw content of file in Git repo over http

```bash
curl http://localhost:5000/raw/helix/test/main/README.md
```

## Git Protocols and APIs

### 1. Git Raw Protocol

Serving content of a file in a git repo.

The requested file is specified by:

* `{owner}`: GitHub organization or user
* `{repo}`: repository name
* `{ref}`: Git reference
  * branch name (e.g. `main`)
  * tag name (e.g. `v1.0`)
  * (full or shorthand) commit id (e.g. `7aeff3d`)

GitHub URLs:

* `https://raw.githubusercontent.com/{owner}/{repo}/{ref}/path/to/file`
* `https://github.com/{owner}/{repo}/raw/{ref}/path/to/file`

Local `git-server` URLs:

* `http://localhost:{port}/raw/{owner}/{repo}/{ref}/path/to/file`
* `http://localhost:{port}/{owner}/{repo}/raw/{ref}/path/to/file`
* `http://raw.localtest.me:{port}/{owner}/{repo}/{ref}/path/to/file` (using wildcarded DNS domain resolving to `127.0.0.1`)

Remote examples:

* `https://raw.githubusercontent.com/adobe/git-server/main/README.md`
* `https://github.com/adobe/git-server/raw/main/README.md`

Local examples:

* `http://raw.localtest.me:5000/adobe/git-server/main/README.md`
* `http://localhost:5000/adobe/git-server/raw/main/README.md`
* `http://localhost:5000/raw/adobe/git-server/main/README.md`

`raw.githubusercontent.com` serves certain file types (e.g. JavaScript, CSS, HTML) with incorrect `Content-Type: text/plain` header. 3rd party solutions like `rawgit.com` address this issue. `git-server` serves files with correct content type.

### 2. Git HTTP Transfer Protocols

Support for `git clone, push, fetch`

Documentation:

* [10.6 Git Internals - Transfer Protocols](https://git-scm.com/book/en/v2/Git-Internals-Transfer-Protocols)
* [Git HTTP transport protocol documentation](https://github.com/git/git/blob/master/Documentation/technical/http-protocol.txt)

#### `git push` support

The served local repo needs to be either a *bare* repo (`git clone --bare` or `git init --bare`) or the following option needs to be set:

```bash
git config receive.denyCurrentBranch updateInstead
```

[more information](https://stackoverflow.com/questions/804545/what-is-this-git-warning-message-when-pushing-changes-to-a-remote-repository/28262104#28262104)

### 3. GitHub API v3

Documentation: [GitHub API v3](https://developer.github.com/v3/)

GitHub Endpoint: `https://api.github.com/`

Local endpoint: `http://localhost:{port}/api` or e.g. `http://api.localtest.me:{port}` (using wildcarded DNS domain resolving to `127.0.0.1`)

Only a small subset of the GitHub API will be supported:

* [Get contents](https://developer.github.com/v3/repos/contents/#get-contents)

  _Note:_ The GitHub implementation supports files up to 1 megabyte in size.

* [Get blob](https://developer.github.com/v3/git/blobs/#get-a-blob)

  _Note:_ The GitHub implementation supports blobs up to 100 megabytes in size.

* [Get tree](https://developer.github.com/v3/git/trees/#get-a-tree-recursively)

* [Get commits](https://developer.github.com/v3/repos/commits/#list-commits-on-a-repository)

* [Get archive link](https://developer.github.com/v3/repos/contents/#get-archive-link)

  _Note:_  This method returns a `302` to a URL to download a tarball or zipball archive for a repository. `git-server` also supports an unofficial `https://codeload.github.com` endpoint that is not rate limited and that doesn't redirect:

  * `https://codeload.github.com/{owner}/{repo}/[zip|tar.gz]/main`

  Related issue/discussion: [#5 Support codeload.github.com](https://github.com/adobe/git-server/issues/5#issuecomment-403072428)

Local examples:

* `http://api.localtest.me:5000/repos/adobe/git-server/contents/README.md?ref=main`
* `http://api.localtest.me:5000/repos/adobe/project-helix/git/blobs/bf13fe66cbee379db6a3e4ebf0300b8bbc0f01b7`
* `http://localhost:5000/api/repos/adobe/git-server/contents/README.md?ref=main`
* `http://localhost:5000/api/repos/adobe/project-helix/git/blobs/bf13fe66cbee379db6a3e4ebf0300b8bbc0f01b7`

### 4. GitHub-like Web Server

_(not yet implemented)_

Endpoint: `https://github.com/`

e.g.

  `https://github.com/{owner}/{repo}`,
  `https://github.com/{owner}/{repo}/blob/{branch}/path/to/file`


CHANGELOG.md
## [1.3.29](https://github.com/adobe/git-server/compare/v1.3.28...v1.3.29) (2021-09-20)


### Bug Fixes

* **deps:** update dependency @adobe/helix-log to v5.0.5 ([96f3356](https://github.com/adobe/git-server/commit/96f3356860e98b1d826d8370e0a2d5f0900520b9))

## [1.3.28](https://github.com/adobe/git-server/compare/v1.3.27...v1.3.28) (2021-09-19)


### Bug Fixes

* **deps:** update dependency http-terminator to v3.0.3 ([9427a17](https://github.com/adobe/git-server/commit/9427a17e42817367eb3b1ca9da899d1394245473))

## [1.3.27](https://github.com/adobe/git-server/compare/v1.3.26...v1.3.27) (2021-09-02)


### Bug Fixes

* **deps:** update dependency @adobe/helix-log to v5.0.4 ([01dc47f](https://github.com/adobe/git-server/commit/01dc47fa8ce22df00585b3947d4a0e523216ac54))

## [1.3.26](https://github.com/adobe/git-server/compare/v1.3.25...v1.3.26) (2021-08-21)


### Bug Fixes

* **deps:** update dependency isomorphic-git to v1.10.0 ([770b9bb](https://github.com/adobe/git-server/commit/770b9bb1b041688530b8f815d3a5844340ad0154))

## [1.3.25](https://github.com/adobe/git-server/compare/v1.3.24...v1.3.25) (2021-08-07)


### Bug Fixes

* **deps:** update dependency async to v3.2.1 ([aaf0815](https://github.com/adobe/git-server/commit/aaf0815c04b69e80049d95733fa719a4464e7332))

## [1.3.24](https://github.com/adobe/git-server/compare/v1.3.23...v1.3.24) (2021-07-17)


### Bug Fixes

* **deps:** update dependency isomorphic-git to v1.9.2 ([94aa43e](https://github.com/adobe/git-server/commit/94aa43e4140fa0d1da3c34fc3ab1ca3e024b0d35))

## [1.3.23](https://github.com/adobe/git-server/compare/v1.3.22...v1.3.23) (2021-07-10)


### Bug Fixes

* **deps:** update dependency isomorphic-git to v1.9.1 ([59e2ddd](https://github.com/adobe/git-server/commit/59e2ddd08606d4fbafbbeea01af359bda7e0c8eb))

## [1.3.22](https://github.com/adobe/git-server/compare/v1.3.21...v1.3.22) (2021-06-26)


### Bug Fixes

* **deps:** update dependency isomorphic-git to v1.8.3 ([a2dfced](https://github.com/adobe/git-server/commit/a2dfced835fbd8e8cc71d708cc00cff65816e135))

## [1.3.21](https://github.com/adobe/git-server/compare/v1.3.20...v1.3.21) (2021-05-15)


### Bug Fixes

* address lint complaints ([97dce0c](https://github.com/adobe/git-server/commit/97dce0cd3661f9ee65c26b3f1315234b15a71d30))

## [1.3.20](https://github.com/adobe/git-server/compare/v1.3.19...v1.3.20) (2021-05-10)


### Bug Fixes

* **deps:** update dependency fs-extra to v10 ([1043559](https://github.com/adobe/git-server/commit/10435596eeeefcbc4ec9eb5e5f33fdbde6a90f41))

## [1.3.19](https://github.com/adobe/git-server/compare/v1.3.18...v1.3.19) (2021-04-17)


### Bug Fixes

* **deps:** update dependency isomorphic-git to v1.8.2 ([13cede4](https://github.com/adobe/git-server/commit/13cede4f9a5c819818bfcf7dac89cfc8076fcd26))

## [1.3.18](https://github.com/adobe/git-server/compare/v1.3.17...v1.3.18) (2021-04-12)


### Bug Fixes

* **deps:** update dependency @adobe/helix-log to v5.0.3 ([#283](https://github.com/adobe/git-server/issues/283)) ([3b2b24a](https://github.com/adobe/git-server/commit/3b2b24ad60cc0c46a771aeea218d5f3828b36998))

## [1.3.17](https://github.com/adobe/git-server/compare/v1.3.16...v1.3.17) (2021-04-02)


### Bug Fixes

* **deps:** update dependency @adobe/helix-log to v5.0.2 ([406471f](https://github.com/adobe/git-server/commit/406471fa299e76c1c2a6675d3e589607837bf39c))

## [1.3.16](https://github.com/adobe/git-server/compare/v1.3.15...v1.3.16) (2021-04-02)


### Bug Fixes

* **deps:** update dependency @adobe/helix-log to v5.0.1 ([9d6770b](https://github.com/adobe/git-server/commit/9d6770bf458750d3674f39e54c591b0db728fd23))

## [1.3.15](https://github.com/adobe/git-server/compare/v1.3.14...v1.3.15) (2021-03-13)


### Bug Fixes

* **deps:** update external fixes ([d9e7537](https://github.com/adobe/git-server/commit/d9e7537b7aa2607d6dba20c599d6d5db00f27105))

## [1.3.14](https://github.com/adobe/git-server/compare/v1.3.13...v1.3.14) (2021-02-25)


### Bug Fixes

* **deps:** update dependency @adobe/helix-log to v5 ([#267](https://github.com/adobe/git-server/issues/267)) ([259c9da](https://github.com/adobe/git-server/commit/259c9daf49448bef2a8c30819fb23f62c1978b39))

## [1.3.13](https://github.com/adobe/git-server/compare/v1.3.12...v1.3.13) (2021-02-20)


### Bug Fixes

* **deps:** update dependency lodash to v4.17.21 ([ddc313a](https://github.com/adobe/git-server/commit/ddc313a21f9b85930dc8a4573a761cbdfca41153))

## [1.3.12](https://github.com/adobe/git-server/compare/v1.3.11...v1.3.12) (2021-02-20)


### Bug Fixes

* **deps:** update dependency mime to v2.5.2 ([8dd1ff8](https://github.com/adobe/git-server/commit/8dd1ff84824c76879029c1787ecb159f1b763606))

## [1.3.11](https://github.com/adobe/git-server/compare/v1.3.10...v1.3.11) (2021-02-15)


### Bug Fixes

* **deps:** update dependency http-terminator to v3 ([1362a58](https://github.com/adobe/git-server/commit/1362a5826b6531e989c6ac68f66627dfb89c7b92))

## [1.3.10](https://github.com/adobe/git-server/compare/v1.3.9...v1.3.10) (2021-01-23)


### Bug Fixes

* **deps:** update external fixes ([28cdfd2](https://github.com/adobe/git-server/commit/28cdfd2879884e6497041dc306f7042d56586d90))

## [1.3.9](https://github.com/adobe/git-server/compare/v1.3.8...v1.3.9) (2021-01-09)


### Bug Fixes

* **deps:** update dependency archiver to v5.2.0 ([af51b9c](https://github.com/adobe/git-server/commit/af51b9c26cd24bb62b039aa265ead20f48123f1c))

## [1.3.8](https://github.com/adobe/git-server/compare/v1.3.7...v1.3.8) (2020-11-18)


### Bug Fixes

* **deps:** update dependency @adobe/helix-log to v4.5.3 ([6c13c93](https://github.com/adobe/git-server/commit/6c13c939d8ca7f452b89aa2c4cc2daba0436ed7b))

## [1.3.7](https://github.com/adobe/git-server/compare/v1.3.6...v1.3.7) (2020-11-09)


### Bug Fixes


config.js
/*
 * Copyright 2018 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

'use strict';

module.exports = {
  appTitle: 'Helix Git Server',
  repoRoot: './repos',
  // repository mapping. allows to 'mount' repositories outside the 'repoRoot' structure.
  virtualRepos: {
    demoOwner: {
      demoRepo: {
        path: './virtual/example',
      },
    },
  },
  listen: {
    http: {
      port: 5000,
      host: '0.0.0.0',
    },
    /*
    // https is optional
    https: {
      // cert: if no file is specfied a selfsigned certificate will be generated on-the-fly
      // cert: './localhost.crt',
      // key: if no file is specfied a key will be generated on-the-fly
      // key: './localhost.key',
      port: 5443,
      host: '0.0.0.0',
    },
    */
  },
  subdomainMapping: {
    // if enabled, <subdomain>.<baseDomain>/foo/bar/baz will be
    // resolved/mapped to 127.0.0.1/<subdomain>/foo/bar/baz
    enable: true,
    baseDomains: [
      // some wildcarded DNS domains resolving to 127.0.0.1
      'localtest.me',
      'lvh.me',
      'vcap.me',
      'lacolhost.com',
    ],
  },
  logs: {
    level: 'info', // fatal, error, warn, info, verbose, debug, trace
    logsDir: './logs',
    reqLogFormat: 'short', // used for morgan (request logging)
  },
};
